### PR TITLE
Handle constant expressions 

### PR DESCRIPTION
**Pull request by Rekt3421**
Adds in functionality to handle more types of constant expressions including arrays and vectors which may have them hiding inside their contents.

### CODE REVIEW

File: lib/BitcastUtils.cpp [hunk 1]
```diff
@@ -815,10 +815,13 @@ void ConvertInto(Type *Ty, IRBuilder<> &Builder,
 
 bool RemoveCstExprFromFunction(Function *F) {
   SmallVector<std::pair<Instruction *, unsigned>, 16> WorkList;
-
```
Comment (rjodinchr) [inline lib/BitcastUtils.cpp:818]: Please avoid unneeded changes like removing an empty line

File: lib/BitcastUtils.cpp [hunk 2]
```diff
@@ -815,10 +815,13 @@ void ConvertInto(Type *Ty, IRBuilder<> &Builder,
...
   auto CheckInstruction = [&WorkList](Instruction *I) {
     for (unsigned OperandId = 0; OperandId < I->getNumOperands(); OperandId++) {
-      if (isa<ConstantExpr>(I->getOperand(OperandId))) {
+      if (dyn_cast<ConstantExpr>(I->getOperand(OperandId))) {
```
Comment (rjodinchr) [inline lib/BitcastUtils.cpp:820]: Can we just have a OR instead of having the same things in every if cases?

File: lib/BitcastUtils.cpp [hunk 3]
```diff
@@ -831,25 +834,53 @@ bool RemoveCstExprFromFunction(Function *F) {
   }
 
   bool Changed = !WorkList.empty();
-
```
Comment (rjodinchr) [inline lib/BitcastUtils.cpp:834]: avoid removing empty line

File: lib/BitcastUtils.cpp [hunk 4]
```diff
@@ -831,25 +834,53 @@ bool RemoveCstExprFromFunction(Function *F) {
...
+        Value *Scalar = B.CreateExtractValue(CstStruct, i);
+        auto *ScalarInst =
+            dyn_cast<ConstantExpr>(Scalar)->getAsInstruction(InsertBefore);
+        if (ScalarInst) {
+          ArrayNew = B.CreateInsertValue(ArrayNew, ScalarInst, i);
+          WorkList.push_back(std::make_pair(ScalarInst, OperandId));
+        }
+      }
+      I->setOperand(OperandId, ArrayNew);
+    }
```
Comment (rjodinchr) [inline lib/BitcastUtils.cpp:862]: else if?

File: lib/BitcastUtils.cpp [hunk 5]
```diff
@@ -831,25 +834,53 @@ bool RemoveCstExprFromFunction(Function *F) {
...
+      for (unsigned i = 0; i < numEle; ++i) {
+        Value *Scalar = B.CreateExtractElement(CstStruct, i);
+        if (auto *CstScalar = dyn_cast<ConstantExpr>(Scalar)) {
+          auto ScalarInst = CstScalar->getAsInstruction(InsertBefore);
+          VecNew = B.CreateInsertElement(VecNew, ScalarInst, i);
+          WorkList.push_back(std::make_pair(ScalarInst, OperandId));
+        }
+      }
+      I->setOperand(OperandId, VecNew);
+    }
```
Comment (rjodinchr) [inline lib/BitcastUtils.cpp:876]: else if?

File: lib/BitcastUtils.cpp [hunk 6]
```diff
@@ -831,25 +834,53 @@ bool RemoveCstExprFromFunction(Function *F) {
...
-                       ->getAsInstruction(InsertBefore);
-    CheckInstruction(Operand);
-    I->setOperand(OperandId, Operand);
-  }
+    if (auto CstStruct = dyn_cast<ConstantArray>(I->getOperand(OperandId))) {
+      unsigned numEle = CstStruct->getType()->getArrayNumElements();
+      Value *ArrayNew = UndefValue::get(CstStruct->getType());
+      for (unsigned i = 0; i < numEle; ++i) {
+        Value *Scalar = B.CreateExtractValue(CstStruct, i);
+        auto *ScalarInst =
```
Comment (rjodinchr) [inline lib/BitcastUtils.cpp:854]: You need to test the result of `dyn_cast<ConstantExpr>` before trying to call `getAsInstruction` as the result of `dyn_cast` can be `nullptr`.
Also even if the cast does not work, something needs to be inserted in the array.

File: lib/BitcastUtils.cpp [hunk 7]
```diff
@@ -831,25 +834,53 @@ bool RemoveCstExprFromFunction(Function *F) {
...
+    }
+    if (auto CstStruct = dyn_cast<ConstantVector>(I->getOperand(OperandId))) {
+      auto FxVecTy = dyn_cast<FixedVectorType>(CstStruct->getType());
+      unsigned numEle = FxVecTy->getNumElements();
+      Value *VecNew = UndefValue::get(CstStruct->getType());
+      for (unsigned i = 0; i < numEle; ++i) {
+        Value *Scalar = B.CreateExtractElement(CstStruct, i);
+        if (auto *CstScalar = dyn_cast<ConstantExpr>(Scalar)) {
+          auto ScalarInst = CstScalar->getAsInstruction(InsertBefore);
+          VecNew = B.CreateInsertElement(VecNew, ScalarInst, i);
```
Comment (rjodinchr) [inline lib/BitcastUtils.cpp:871]: Even if the cast does not work, something needs to be inserted in the array.

File: lib/LogicalPointerToIntPass.cpp [hunk 8]
```diff
@@ -15,6 +15,7 @@
 #include "llvm/Analysis/ScalarEvolution.h"
 #include "llvm/IR/Constants.h"
 #include "llvm/IR/Instructions.h"
+#include "llvm/IR/Type.h"
```
Comment (rjodinchr) [inline lib/LogicalPointerToIntPass.cpp:18]: Please avoid unneeded changes

File: lib/SPIRVProducerPass.cpp [hunk 9]
```diff
@@ -399,7 +399,9 @@ struct SPIRVProducerPassImpl {
   SmallPtrSet<Value *, 16> &getGlobalConstArgSet() {
     return GlobalConstArgumentSet;
   }
-  StrideTypeList &getTypesNeedingArrayStride() { return TypesNeedingArrayStride; }
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:402]: I think this is comming from clang-format, but as you do not modify it, you should not have to commit that. Please try to revert that.

File: lib/SPIRVProducerPass.cpp [hunk 10]
```diff
@@ -440,7 +442,6 @@ struct SPIRVProducerPassImpl {
   SPIRVID getSPIRVType(Type *Ty);
   SPIRVID getSPIRVConstant(Constant *Cst);
   SPIRVID getSPIRVInt32Constant(uint32_t CstVal);
-  SPIRVID getSPIRVInt64Constant(uint64_t CstVal);
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:443]: Why do you want to remove this function?

File: lib/SPIRVProducerPass.cpp [hunk 11]
```diff
@@ -831,6 +832,7 @@ PreservedAnalyses SPIRVProducerPass::run(Module &M,
 }
 
 bool SPIRVProducerPassImpl::runOnModule(Module &M) {
+
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:835]: empty line

File: lib/SPIRVProducerPass.cpp [hunk 12]
```diff
@@ -1285,8 +1288,7 @@ void SPIRVProducerPassImpl::FindTypesForResourceVars() {
       const spv::BuiltIn BuiltinType = GetBuiltin(GV.getName());
       if (module_scope_constant_external_init &&
           spv::BuiltInMax == BuiltinType) {
-        StructTypesNeedingBlock.insert(
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:1288]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 13]
```diff
@@ -1843,22 +1845,22 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVType(Type *Ty, bool needs_layout) {
       const auto dim = ImageDimensionality(ext_ty);
       const auto sampled = IsSampledImageType(ext_ty);
       switch (dim) {
-        case spv::Dim1D:
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:1846]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 14]
```diff
@@ -1877,30 +1879,30 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVType(Type *Ty, bool needs_layout) {
       SPIRVID SampledTyID;
       // None of the sampled types have a layout.
       if (IsFloatImageType(ext_ty)) {
-          SampledTyID =
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:1880]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 15]
```diff
@@ -2350,7 +2356,7 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVConstant(Constant *C) {
     }
   } else if (Cst->isNullValue()) {
     Opcode = spv::OpConstantNull;
-  } else if(const Function *Fn = dyn_cast<Function>(Cst)) {
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:2353]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 16]
```diff
@@ -1936,10 +1938,10 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVType(Type *Ty, bool needs_layout) {
       // Only need a sampled version of the type if it is used with a sampler.
       // In SPIR-V 1.6 or later, sampled image dimension must not be Buffer
       if (Sampled == 1 && ImageDimensionality(ext_ty) != spv::DimBuffer) {
-          Ops.clear();
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:1939]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 17]
```diff
@@ -1925,7 +1927,7 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVType(Type *Ty, bool needs_layout) {
       // 2 indicates will be used without a sampler (a storage image)
       uint32_t Sampled = 1;
       if (!IsSampledImageType(ext_ty)) {
-          Sampled = 2;
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:1928]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 18]
```diff
@@ -2682,8 +2688,7 @@ void SPIRVProducerPassImpl::GenerateGlobalVar(GlobalVariable &GV) {
         SPIRVID ZDimCstID =
             getSPIRVValue(mdconst::extract<ConstantInt>(MD->getOperand(2)));
 
-        Ops << GV.getValueType() << XDimCstID << YDimCstID
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:2685]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 19]
```diff
@@ -3505,7 +3510,7 @@ spv::Op SPIRVProducerPassImpl::GetSPIRVCastOpcode(Instruction &I) {
       {Instruction::BitCast, spv::OpBitcast},
       {Instruction::PtrToInt, spv::OpConvertPtrToU},
       {Instruction::IntToPtr, spv::OpConvertUToPtr},
-      };
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:3508]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 20]
```diff
@@ -3931,8 +3936,7 @@ SPIRVProducerPassImpl::GenerateImageInstruction(CallInst *Call,
 
     Value *Image = Call->getArgOperand(0);
     const uint32_t dim = ImageNumDimensions(image_ty);
-    const uint32_t components =
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:3934]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 21]
```diff
@@ -7019,7 +7023,8 @@ void SPIRVProducerPassImpl::GenerateKernelReflection() {
 
     // Generate the reflection for the image channel getter function if it is
     // used in this kernel.
-    auto *image_getter_md = F.getMetadata(clspv::PushConstantsMetadataImageChannelName());
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:7022]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 22]
```diff
@@ -7100,10 +7105,9 @@ void SPIRVProducerPassImpl::AddArgumentReflection(
 
   if (clspv::Option::KernelArgInfo()) {
     assert(kernelFn.getMetadata("kernel_arg_type") &&
-            kernelFn.getMetadata("kernel_arg_addr_space") &&
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:7103]: clang-format, not needed, please revert

File: lib/SPIRVProducerPass.cpp [hunk 23]
```diff
@@ -7161,7 +7165,6 @@ void SPIRVProducerPassImpl::AddArgumentReflection(
     }
     auto type_qual_enum = getSPIRVInt32Constant(type_qual_enum_value);
     Ops << type_qual_enum;
-
```
Comment (rjodinchr) [inline lib/SPIRVProducerPass.cpp:7164]: clang-format, not needed, please revert

File: test/Structs/insert_value_issue_14.cl [hunk 24]
```diff
@@ -28,19 +28,18 @@ kernel void foo(global S* A, global uchar4* B, int n) {
...
+// CHECK-DAG: [[undef_uint:%[a-zA-Z0-9_]+]] = OpUndef [[uint]]
 // no longer checked: [[undef_struct:%[_a-zA-Z0-9]+]] = OpUndef [[struct]]
 
+// CHECK: [[loadOp:%[a-zA-z0-9_]+]] = OpLoad [[struct]] %{{.*}}
+// CHECK: [[theconst:%[a-zA-Z0-9_]+]] = OpCompositeExtract [[uint]] [[loadOp]] 0
 // CHECK: [[and:%[a-zA-Z0-9_]+]] = OpBitwiseAnd [[uint]] %{{.*}} [[int_255]]
 // CHECK: [[mask_255:%[a-zA-Z0-9_]+]] = OpShiftLeftLogical %uint [[int_255]] [[int_0]]
 // CHECK: [[mask:%[a-zA-Z0-9_]+]] = OpNot %uint [[mask_255]]
-// CHECK: [[otherelems:%[a-zA-Z0-9_]+]] = OpBitwiseAnd %uint [[theconst]] [[mask]]
+// CHECK: [[otherelems:%[a-zA-Z0-9_]+]] = OpBitwiseAnd %uint [[undef_uint]] [[mask]]
```
Comment (rjodinchr) [inline test/Structs/insert_value_issue_14.cl:43]: BitwiseAnd with `Undef` does not feel right, not sure what is happening here

File: test/char4_insert.cl [hunk 25]
```diff
@@ -13,7 +13,7 @@ kernel void foo(global uchar4* A, int n) {
 // CHECK:  [[_uint_0:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 0
 // CHECK:  [[_uint_255:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 255
 // CHECK:  [[_uint_16:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 16
-// CHECK:  [[_uint_16908292:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 16908292
+// CHECK:  [[_uint_16908292:%[0-9a-zA-Z_]+]] = OpUndef [[_uint]]
```
Comment (rjodinchr) [inline test/char4_insert.cl:16]: This node is used to perform a bitwiseAnd, it does not feel right.

File: test/char4_insert_from_float.cl [hunk 26]
```diff
@@ -16,7 +16,7 @@ kernel void foo(global char4* A, float f) {
 // CHECK:  [[_uint_0:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 0
 // CHECK:  [[_uint_255:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 255
 // CHECK:  [[_uint_16:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 16
-// CHECK:  [[_uint_16908292:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 16908292
+// CHECK:  [[_uint_16908292:%[0-9a-zA-Z_]+]] = OpUndef [[_uint]]
```
Comment (rjodinchr) [inline test/char4_insert_from_float.cl:19]: This node is used to perform a bitwiseAnd, it does not feel right.

File: test/uchar4_insert_from_float.cl [hunk 27]
```diff
@@ -20,7 +20,7 @@ kernel void foo(global uchar4* A, float f) {
 // CHECK:  [[_uint_0:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 0
 // CHECK:  [[_uint_255:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 255
 // CHECK:  [[_uint_16:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 16
-// CHECK:  [[_uint_16908292:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 16908292
+// CHECK:  [[_uint_16908292:%[0-9a-zA-Z_]+]] = OpUndef [[_uint]]
```
Comment (rjodinchr) [inline test/uchar4_insert_from_float.cl:23]: This node is used to perform a bitwiseAnd, it does not feel right.

File: test/LongVectorLowering/constant_vector.ll [hunk 28]
```diff
@@ -29,7 +29,7 @@ entry:
 ; CHECK: [[ins6:%[^ ]+]] = insertvalue [16 x float] [[ins5]], float [[ld6]], 6
 ; CHECK: [[ld7:%[^ ]+]] = extractvalue [8 x float] [[load]], 7
 ; CHECK: [[ins7:%[^ ]+]] = insertvalue [16 x float] [[ins6]], float [[ld7]], 7
-; CHECK: [[ins8:%[^ ]+]] = insertvalue [16 x float] [[ins7]], float 0.000000e+00, 8
+; CHECK: [[ins8:%[^ ]+]] = insertvalue [16 x float] [[ins7]], float undef, 8
```
Comment (rjodinchr) [inline test/LongVectorLowering/constant_vector.ll:32]: inserting undef does not feel right

**Review by rjodinchr — CHANGES_REQUESTED** [review]
Please avoid committing clang-format changes of line you did not touch. It makes the review longer for no reason.
You did not add a specific test for what you implemented, please add one.
Some tests have undefs that do not feel correct (inserted in a vector, or used for bitcast). While it is normal to have a undef vector/array that you can then populate, using undefined values for such operations should not happen I believe. Might be related to not checking if the dyn_cast worked for ConstantArray.

File: lib/BitcastUtils.cpp (hunk 1 above)
Comment (Rekt3421) [reply lib/BitcastUtils.cpp:818]: Done

File: lib/SPIRVProducerPass.cpp (hunk 10 above)
Comment (Rekt3421) [reply lib/SPIRVProducerPass.cpp:443]: I thought I added it in myself for some reason

File: test/Structs/insert_value_issue_14.cl [hunk 29]
```diff
@@ -28,7 +28,6 @@ kernel void foo(global S* A, global uchar4* B, int n) {
 
 // CHECK-DAG: [[uint:%[_a-zA-Z0-9]+]] = OpTypeInt 32 0
 // no longer checked: [[struct:%[_a-zA-Z0-9]+]] = OpTypeStruct [[uint]] [[uint]] [[uint]]
-
```
Comment (rjodinchr) [inline test/Structs/insert_value_issue_14.cl:31]: please revert

**Review by rjodinchr — APPROVED** [review]
Looking great! please fix the issue in `CI-check-format` and that should be ready to land

File: lib/BitcastUtils.cpp [hunk 30]
```diff
@@ -837,19 +839,53 @@ bool RemoveCstExprFromFunction(Function *F) {
...
       InsertBefore =
           phi->getIncomingBlock(phi->getIncomingValueNumForOperand(OperandId))
               ->getFirstNonPHI();
     }
 
-    auto Operand = dyn_cast<ConstantExpr>(I->getOperand(OperandId))
-                       ->getAsInstruction(InsertBefore);
-    CheckInstruction(Operand);
-    I->setOperand(OperandId, Operand);
+    if (auto CstStruct = dyn_cast<ConstantArray>(I->getOperand(OperandId))) {
```
Comment (alan-baker) [inline lib/BitcastUtils.cpp:850]: Not a great name choice.

```suggestion
    if (auto CstArray = dyn_cast<ConstantArray>(I->getOperand(OperandId))) {
```

File: lib/BitcastUtils.cpp [hunk 31]
```diff
@@ -837,19 +839,53 @@ bool RemoveCstExprFromFunction(Function *F) {
...
+          auto *ScalarInst = ScalarCst->getAsInstruction(InsertBefore);
+          ArrayNew = B.CreateInsertValue(ArrayNew, ScalarInst, i);
+          WorkList.push_back(std::make_pair(ScalarInst, OperandId));
+        } else {
+          ArrayNew = B.CreateInsertValue(ArrayNew, Scalar, i);
+        }
+      }
+      I->setOperand(OperandId, ArrayNew);
+    }
+    else if (auto CstStruct = dyn_cast<ConstantVector>(I->getOperand(OperandId))) {
```
Comment (alan-baker) [inline lib/BitcastUtils.cpp:866]:
```suggestion
    else if (auto CstVector = dyn_cast<ConstantVector>(I->getOperand(OperandId))) {
```

File: lib/SPIRVProducerPass.cpp [hunk 32]
```diff
@@ -2286,9 +2287,13 @@ SPIRVID SPIRVProducerPassImpl::getSPIRVConstant(Constant *C) {
     Ops << LiteralNum;
   } else if (isa<ConstantDataSequential>(Cst) &&
              cast<ConstantDataSequential>(Cst)->isString()) {
-    Cst->print(errs());
-    llvm_unreachable("Implement this Constant");
 
+    const ConstantDataSequential *CDS = dyn_cast<ConstantDataSequential>(Cst);
```
Comment (alan-baker) [inline lib/SPIRVProducerPass.cpp:2291]: I don't see tests for this.

File: lib/BitcastUtils.cpp [hunk 33]
```diff
@@ -818,8 +818,10 @@ bool RemoveCstExprFromFunction(Function *F) {
 
   auto CheckInstruction = [&WorkList](Instruction *I) {
     for (unsigned OperandId = 0; OperandId < I->getNumOperands(); OperandId++) {
-      if (isa<ConstantExpr>(I->getOperand(OperandId))) {
-        WorkList.push_back(std::make_pair(I, OperandId));
+      if (dyn_cast<ConstantExpr>(I->getOperand(OperandId)) ||
+          dyn_cast<ConstantVector>(I->getOperand(OperandId)) || 
+          dyn_cast<ConstantArray>(I->getOperand(OperandId))) {
```
Comment (alan-baker) [inline lib/BitcastUtils.cpp:823]: This adds constant data sequential handling. Is constant aggregate handling also needed?

**Review by alan-baker — CHANGES_REQUESTED** [review]

File: lib/BitcastUtils.cpp (hunk 33 above)
Comment (rjodinchr) [reply lib/BitcastUtils.cpp:823]: I guess, it won't be part of this PR. It was not part of the issues in #1223.

**Review by rjodinchr — APPROVED** [review]

**Review by alan-baker — APPROVED** [review]

**Pull request merged**